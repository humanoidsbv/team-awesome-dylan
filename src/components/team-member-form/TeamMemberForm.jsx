import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';

import styles from './TeamMemberForm.module.css';

function TeamMemberForm({
  createTeamMember,
  toggleTeamMemberForm,
  isTeamMemberFormVisible
}) {
  const today = new Date()
    .toISOString()
    .split('T')
    .shift();

  const [address, setAddress] = useState('');
  const [bio, setBio] = useState('');
  const [currentClient, setCurrentClient] = useState('To be confirmed');
  const [emailAddress, setEmailAddress] = useState('');
  const [employeeNumber, setEmployeeNumber] = useState('To be confirmed');
  const [employeeFunction, setEmployeeFunction] = useState('To be confirmed');
  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');
  const [locality, setLocality] = useState('');
  const [postalCode, setPostalCode] = useState('');
  const [socialProfileOne, setSocialProfileOne] = useState('');
  const [socialProfileTwo, setSocialProfileTwo] = useState('');
  const [startingDate, setStartingDate] = useState(today);
  const [validity, setValidity] = useState({});

  const formRef = useRef(null);

  const handleBlur = event => {
    setValidity({
      ...validity,
      form: formRef.current.checkValidity(),
      [event.target.name]: event.target.checkValidity()
    });
  };

  const handleCancel = event => {
    event.preventDefault();

    setAddress('');
    setBio('');
    setEmailAddress('');
    setFirstName('');
    setEmployeeFunction('');
    setLastName('');
    setLocality('');
    setPostalCode('');
    setSocialProfileOne('');
    setSocialProfileTwo('');
    setStartingDate('');
    toggleTeamMemberForm();
    setEmployeeNumber('');
    setCurrentClient('');
  };

  const handleSubmit = event => {
    event.preventDefault();

    setAddress('');
    setBio('');
    setEmailAddress('');
    setEmployeeFunction('');
    setFirstName('');
    setLastName('');
    setLocality('');
    setPostalCode('');
    setSocialProfileOne('');
    setSocialProfileTwo('');
    setStartingDate('');
    setEmployeeNumber('');
    setCurrentClient('');

    createTeamMember({
      address,
      bio,
      currentClient,
      emailAddress,
      employeeNumber,
      employeeFunction,
      firstName,
      id: Math.random(),
      lastName,
      locality,
      postalCode,
      socialProfileOne,
      socialProfileTwo,
      startingDate
    });
    toggleTeamMemberForm();
  };

  return (
    <div
      className={`${styles.teamMemberForm} ${
        isTeamMemberFormVisible
          ? styles.teamMemberForm
          : styles.teamMemberFormHide
      }`}
    >
      <div>
        <div className={styles.headerTeamMemberForm}>
          <span className={styles.headerTeamMemberFormText}>
            Add new team member
          </span>
          <button
            type="button"
            className={styles.cancelButton}
            onClick={handleCancel}
          >
            Cancel
          </button>
          <button
            type="button"
            className={styles.saveButton}
            onClick={handleSubmit}
            disabled={validity.form !== true}
          >
            Save
          </button>
        </div>
      </div>
      <div className={styles.personalDetailsTab}>Personal details</div>
      <form className={styles.form} onSubmit={handleSubmit} ref={formRef}>
        <div className={styles.formColumnOne}>
          <div className={styles.nameInputs}>
            <label className={styles.labelFirstName} htmlFor="firstName">
              First Name
              <input
                className={`${styles.inputFirstNameValid} ${
                  validity.firstName === false
                    ? styles.inputFirstNameInvalid
                    : styles.inputFirstNameValid
                }`}
                id="firstName"
                maxLength="35"
                minLength="2"
                name="firstName"
                onBlur={handleBlur}
                onChange={({ target }) => setFirstName(target.value)}
                required
                value={firstName}
              />
            </label>
            <label className={styles.labelLastName} htmlFor="lastName">
              Last Name
              <input
                className={`${styles.inputLastNameValid} ${
                  validity.lastName === false
                    ? styles.inputLastNameInvalid
                    : styles.inputLastNameValid
                }`}
                id="lastName"
                maxLength="35"
                minLength="2"
                name="lastName"
                onBlur={handleBlur}
                onChange={({ target }) => setLastName(target.value)}
                required
                value={lastName}
              />
            </label>
          </div>
          <label className={styles.labelEmailAddress} htmlFor="emailAddress">
            E-mail Address
            <input
              className={`${styles.inputEmailAddressValid} ${
                validity.emailAddress === false
                  ? styles.inputEmailAddressInvalid
                  : styles.inputEmailAddressNameValid
              }`}
              id="emailAddress"
              maxLength="35"
              minLength="2"
              name="emailAddress"
              onBlur={handleBlur}
              onChange={({ target }) => setEmailAddress(target.value)}
              required
              value={emailAddress}
            />
          </label>
          <label className={styles.labelBio} htmlFor="bio">
            Bio
            <input
              className={`${styles.inputBioValid} ${
                validity.bio === false
                  ? styles.inputBioInvalid
                  : styles.inputBioValid
              }`}
              id="bio"
              maxLength="35"
              minLength="2"
              name="bio"
              onBlur={handleBlur}
              onChange={({ target }) => setBio(target.value)}
              required
              value={bio}
            />
          </label>
        </div>
        <div className={styles.formColumnTwo}>
          <label className={styles.labelAddress} htmlFor="address">
            Address
            <input
              className={`${styles.inputAddressValid} ${
                validity.address === false
                  ? styles.inputAddressInvalid
                  : styles.inputAddressValid
              }`}
              id="address"
              maxLength="35"
              minLength="2"
              name="address"
              onBlur={handleBlur}
              onChange={({ target }) => setAddress(target.value)}
              required
              value={address}
            />
          </label>
          <div className={styles.localityPostalCodeInputs}>
            <label className={styles.labelPostalCode} htmlFor="postalCode">
              ZIP code
              <input
                className={`${styles.inputPostalCodeValid} ${
                  validity.postalCode === false
                    ? styles.inputPostalCodeInvalid
                    : styles.inputPostalCodeValid
                }`}
                id="postalCode"
                maxLength="35"
                minLength="2"
                name="postalCode"
                onBlur={handleBlur}
                onChange={({ target }) => setPostalCode(target.value)}
                required
                value={postalCode}
              />
            </label>
            <label className={styles.labelLocality} htmlFor="locality">
              City
              <input
                className={`${styles.inputLocalityValid} ${
                  validity.locality === false
                    ? styles.inputLocalityInvalid
                    : styles.inputLocalityValid
                }`}
                id="locality"
                maxLength="35"
                minLength="2"
                name="locality"
                onBlur={handleBlur}
                onChange={({ target }) => setLocality(target.value)}
                required
                value={locality}
              />
            </label>
          </div>
          <label
            className={styles.labelSocialProfiles}
            htmlFor="socialProfiles"
          >
            Social Profiles
            <input
              className={`${styles.inputSocialProfileOneValid} ${
                validity.socialProfileOne === false
                  ? styles.inputSocialProfileOneInvalid
                  : styles.inputSocialProfileOneValid
              }`}
              id="socialProfileOne"
              maxLength="35"
              minLength="2"
              name="socialProfileOne"
              onBlur={handleBlur}
              onChange={({ target }) => setSocialProfileOne(target.value)}
              required
              value={socialProfileOne}
            />
            <input
              className={`${styles.inputSocialProfileTwoValid} ${
                validity.socialProfileTwo === false
                  ? styles.inputSocialProfileTwoInvalid
                  : styles.inputSocialProfileTwoValid
              }`}
              id="socialProfileTwo"
              maxLength="35"
              minLength="2"
              name="socialProfileTwo"
              onBlur={handleBlur}
              onChange={({ target }) => setSocialProfileTwo(target.value)}
              required
              value={socialProfileTwo}
            />
          </label>
        </div>
      </form>
    </div>
  );
}

TeamMemberForm.propTypes = {
  createTeamMember: PropTypes.func.isRequired,
  toggleTeamMemberForm: PropTypes.func.isRequired,
  isTeamMemberFormVisible: PropTypes.bool.isRequired
};

export default TeamMemberForm;
