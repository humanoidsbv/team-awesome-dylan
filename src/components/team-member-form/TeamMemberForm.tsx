import React, { useState, useRef } from 'react';

import Input from '../../shared/components/input/Input';
import styles from './TeamMemberForm.module.css';
import { ActionInterface } from '../../redux/team-members/types';
import { ValidityState } from '../../shared/types';

interface TeamMemberFormProps {
  createTeamMember: (newTeamMember: {}) => void;
  isTeamMemberFormVisible: boolean;
  toggleTeamMemberForm: () => void;
}

function TeamMemberForm({
  createTeamMember,
  toggleTeamMemberForm,
  isTeamMemberFormVisible
}: TeamMemberFormProps): React.ReactElement {
  const today = new Date()
    .toISOString()
    .split('T')
    .shift();

  const [address, setAddress] = useState('');
  const [bio, setBio] = useState('');
  const [currentClient, setCurrentClient] = useState('To be confirmed');
  const [emailAddress, setEmailAddress] = useState('');
  const [employeeFunction, setEmployeeFunction] = useState('To be confirmed');
  const [employeeNumber, setEmployeeNumber] = useState('To be confirmed');
  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');
  const [locality, setLocality] = useState('');
  const [postalCode, setPostalCode] = useState('');
  const [socialProfileOne, setSocialProfileOne] = useState('');
  const [socialProfileTwo, setSocialProfileTwo] = useState('');
  const [startingDate, setStartingDate] = useState(today);
  const [validity, setValidity] = useState<ValidityState>({});

  const formRef = useRef(null);

  const handleBlur = (event: React.ChangeEvent<HTMLSelectElement | HTMLInputElement>): void => {
    setValidity({
      ...validity,
      form: formRef.current.checkValidity(),
      [event.target.name]: event.target.checkValidity()
    });
  };

  const resetFormStates = (): void => {
    setAddress('');
    setBio('');
    setEmailAddress('');
    setFirstName('');
    setEmployeeFunction('');
    setLastName('');
    setLocality('');
    setPostalCode('');
    setSocialProfileOne('');
    setSocialProfileTwo('');
    setStartingDate('');
    setEmployeeNumber('');
    setCurrentClient('');
  };

  const handleCancel = (): void => {
    event.preventDefault();
    resetFormStates();
    toggleTeamMemberForm();
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement | HTMLButtonElement>): void => {
    event.preventDefault();
    resetFormStates();

    createTeamMember({
      address,
      bio,
      currentClient,
      emailAddress,
      employeeNumber,
      employeeFunction,
      firstName,
      id: Math.random(),
      lastName,
      locality,
      postalCode,
      socialProfileOne,
      socialProfileTwo,
      startingDate
    });
    toggleTeamMemberForm();
  };

  return (
    <div
      className={`${styles.teamMemberForm} ${
        !isTeamMemberFormVisible ? styles.teamMemberFormHide : ''
      }`}
    >
      <div>
        <div className={styles.headerTeamMemberForm}>
          <span className={styles.headerTeamMemberFormText}>
            Add new team member
          </span>
          <button
            type="button"
            className={styles.cancelButton}
            onClick={handleCancel}
          >
            Cancel
          </button>
          <button
            type="button"
            className={styles.saveButton}
            onClick={handleSubmit}
            disabled={validity.form !== true}
          >
            Save
          </button>
        </div>
      </div>
      <div className={styles.personalDetailsTab}>Personal details</div>
      <form className={styles.form} onSubmit={handleSubmit} ref={formRef}>
        <div className={styles.formColumnOne}>
          <div className={styles.nameInputs}>
            <label className={styles.labelFirstName} htmlFor="firstName">
              First Name
              <Input
                isValid={validity.firstName !== false}
                maxLength={35}
                minLength={2}
                name="firstName"
                onBlur={handleBlur}
                onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setFirstName(target.value)}
                required
                value={firstName}
              />
            </label>
            <label className={styles.labelLastName} htmlFor="lastName">
              Last Name
              <Input
                isValid={validity.lastName !== false}
                maxLength={35}
                minLength={2}
                name="lastName"
                onBlur={handleBlur}
                onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setLastName(target.value)}
                required
                value={lastName}
              />
            </label>
          </div>
          <label className={styles.labelEmailAddress} htmlFor="emailAddress">
            E-mail Address
            <Input
              isValid={validity.emailAddress !== false}
              maxLength={35}
              minLength={2}
              name="emailAddress"
              onBlur={handleBlur}
              onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setEmailAddress(target.value)}
              required
              value={emailAddress}
            />
          </label>
          <label className={styles.labelBio} htmlFor="bio">
            Bio
            <Input
              isMultiLine
              isValid={validity.bio !== false}
              maxLength={35}
              minLength={2}
              name="bio"
              onBlur={handleBlur}
              onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setBio(target.value)}
              required
              value={bio}
            />
          </label>
        </div>
        <div className={styles.formColumnTwo}>
          <label className={styles.labelAddress} htmlFor="address">
            Address
            <Input
              isValid={validity.address !== false}
              maxLength={35}
              minLength={2}
              name="address"
              onBlur={handleBlur}
              onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setAddress(target.value)}
              required
              value={address}
            />
          </label>
          <div className={styles.localityPostalCodeInputs}>
            <label className={styles.labelPostalCode} htmlFor="postalCode">
              Postal code
              <Input
                isValid={validity.postalCode !== false}
                maxLength={35}
                minLength={2}
                name="postalCode"
                onBlur={handleBlur}
                onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setPostalCode(target.value)}
                required
                value={postalCode}
              />
            </label>
            <label className={styles.labelLocality} htmlFor="locality">
              Locality
              <Input
                isValid={validity.locality !== false}
                maxLength={35}
                minLength={2}
                name="locality"
                onBlur={handleBlur}
                onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setLocality(target.value)}
                required
                value={locality}
              />
            </label>
          </div>
          <label
            className={styles.labelSocialProfiles}
            htmlFor="socialProfileOne"
          >
            Social Profiles
            <Input
              isValid={validity.socialProfileOne !== false}
              maxLength={35}
              minLength={2}
              name="socialProfileOne"
              onBlur={handleBlur}
              onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setSocialProfileOne(target.value)}
              required
              value={socialProfileOne}
            />
            <Input
              isValid={validity.socialProfileTwo !== false}
              maxLength={35}
              minLength={2}
              name="socialProfileTwo"
              onBlur={handleBlur}
              onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setSocialProfileTwo(target.value)}
              required
              value={socialProfileTwo}
            />
          </label>
        </div>
      </form>
    </div>
  );
}

export default TeamMemberForm;
