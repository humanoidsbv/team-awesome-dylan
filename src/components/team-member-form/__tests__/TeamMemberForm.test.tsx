import * as React from 'react';
import { shallow } from 'enzyme';

import TeamMemberForm from '../TeamMemberForm';

it('matches snapshot', () => {
  const teamMemberForm = shallow(
    <TeamMemberForm
      isTeamMemberFormVisible={true}
      createTeamMember={({}) => {}}
      toggleTeamMemberForm={() => null}
    />
  );
  expect(teamMemberForm).toMatchSnapshot();
});
