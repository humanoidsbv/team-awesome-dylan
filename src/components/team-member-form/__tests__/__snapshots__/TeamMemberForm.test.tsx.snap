// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="teamMemberForm "
>
  <div>
    <div
      className="headerTeamMemberForm"
    >
      <span
        className="headerTeamMemberFormText"
      >
        Add new team member
      </span>
      <button
        className="cancelButton"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
      <button
        className="saveButton"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="personalDetailsTab"
  >
    Personal details
  </div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="formColumnOne"
    >
      <div
        className="nameInputs"
      >
        <label
          className="labelFirstName"
          htmlFor="firstName"
        >
          First Name
          <styled.input
            isValid={true}
            maxLength={35}
            minLength={2}
            name="firstName"
            onBlur={[Function]}
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
        <label
          className="labelLastName"
          htmlFor="lastName"
        >
          Last Name
          <styled.input
            isValid={true}
            maxLength={35}
            minLength={2}
            name="lastName"
            onBlur={[Function]}
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </div>
      <label
        className="labelEmailAddress"
        htmlFor="emailAddress"
      >
        E-mail Address
        <styled.input
          isValid={true}
          maxLength={35}
          minLength={2}
          name="emailAddress"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          value=""
        />
      </label>
      <label
        className="labelBio"
        htmlFor="bio"
      >
        Bio
        <styled.input
          isMultiLine={true}
          isValid={true}
          maxLength={35}
          minLength={2}
          name="bio"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          value=""
        />
      </label>
    </div>
    <div
      className="formColumnTwo"
    >
      <label
        className="labelAddress"
        htmlFor="address"
      >
        Address
        <styled.input
          isValid={true}
          maxLength={35}
          minLength={2}
          name="address"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          value=""
        />
      </label>
      <div
        className="localityPostalCodeInputs"
      >
        <label
          className="labelPostalCode"
          htmlFor="postalCode"
        >
          Postal code
          <styled.input
            isValid={true}
            maxLength={35}
            minLength={2}
            name="postalCode"
            onBlur={[Function]}
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
        <label
          className="labelLocality"
          htmlFor="locality"
        >
          Locality
          <styled.input
            isValid={true}
            maxLength={35}
            minLength={2}
            name="locality"
            onBlur={[Function]}
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </div>
      <label
        className="labelSocialProfiles"
        htmlFor="socialProfileOne"
      >
        Social Profiles
        <styled.input
          isValid={true}
          maxLength={35}
          minLength={2}
          name="socialProfileOne"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          value=""
        />
        <styled.input
          isValid={true}
          maxLength={35}
          minLength={2}
          name="socialProfileTwo"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          value=""
        />
      </label>
    </div>
  </form>
</div>
`;
